# Data Science Mathematics
# Simple Linear Regression
# In-Class Activity

Refer to your class handout for background information.

In [1]:
import numpy as np
from scipy import stats

Let's instantiate the data set.

In [2]:
submarine_sightings = np.array([1,2,3,4,5,6,7,8,9,10])
cyber_activity_metric = np.array([0.021025,0.022103,0.023237,0.024428,0.025681,0.026997,0.028381,0.029836,0.031366,0.032974])

Now, let's calculate our regression values.

In [3]:
slope, intercept, r_value, p_value, std_err = stats.linregress(submarine_sightings,cyber_activity_metric)

In [5]:
slope, intercept, r_value, p_value, std_err

(0.001324557575757576,
 0.019317733333333337,
 0.9980126947882119,
 6.807697897873586e-11,
 2.9567948236252566e-05)

Next, print the R^2 value.  How good is your fit?

In [4]:
print('r-squared:', r_value**2)

r-squared: 0.9960293389584285


### Question 1: 

#1
a) correlation coefficient = r_value = .998

b) alpha_hat = intercept = .019, beta_hat = slope = .0013

c) as observed from the correlation coefficient being close to 1, these two variables do demonstrate a linear relationship. 

d) based on the linear relationship between the number of submarine sightings and the cyber activity metric, the two observations are correlated. 



In [2]:
step = 0.1
pos = [1,5,26] #f(m,b) = x**b + y**b)
for i in range(5):
    pos[0] = round((pos[0] - (step * 2 * pos[0])),2)
    pos[1] = round((pos[1] - (step * 2 * pos[1])),2)
    pos[2] = round((pos[0]**2 + pos[1]**2),2)
    
    print("Step {}".format(i))
    print (pos)
    print ("\n")

Step 0
[0.8, 4.0, 16.64]


Step 1
[0.64, 3.2, 10.65]


Step 2
[0.51, 2.56, 6.81]


Step 3
[0.41, 2.05, 4.37]


Step 4
[0.33, 1.64, 2.8]




In [3]:
step = 0.5
pos = [1,5,26] #f(m,b) = x**b + y**b)
for i in range(5):
    pos[0] = round((pos[0] - (step * 2 * pos[0])),2)
    pos[1] = round((pos[1] - (step * 2 * pos[1])),2)
    pos[2] = round((pos[0]**2 + pos[1]**2),2)
    
    print("Step {}".format(i))
    print (pos)
    print ("\n")

Step 0
[0.0, 0.0, 0.0]


Step 1
[0.0, 0.0, 0.0]


Step 2
[0.0, 0.0, 0.0]


Step 3
[0.0, 0.0, 0.0]


Step 4
[0.0, 0.0, 0.0]




***Now save your output.  Go to File -> Print Preview and save your final output as a PDF.  Turn in to your Instructor, along with any additional sheets.

#2 
a) with a learning rate of .1 and only 5 steps, the model begins to converge, but does not have enough time arrive at the local minima. 

b) with a learning rate of .5, the model converges immediately due to it's relationship with the partial derivatives of the original loss function. 

c) In this case, the ideal learning rate is .5. This may not always be the case. A learning rate that is too large may miss the local optima, while a learning rate that is too small may take too long to converge on the local optima, as demonstrated in part a) of this problem. 